### PR TITLE
Run expression tests when new functions are added

### PR DESCRIPTION
**Pull request by serban-nicusor-toptal**
## Release notes

Run expression tests when new functions are added in Jenkins pipeline.

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

#836

### CODE REVIEW

**Comment by serban-nicusor-toptal** [comment]
Hey @rok-cesnovar I'm not sure where to copy the binary since `bin/stanc` is in cmdstan. I hope I got everything else correct.
Let me know if I should make any change,  thanks!

File: Jenkinsfile [hunk 1]
```diff
@@ -342,6 +355,40 @@ pipeline {
...
+            when {
+                expression {
+                    !skipExpressionTests
+                }
+            }
+            agent any
+            steps {
+
+                sh """
+                    git clone --recursive https://github.com/stan-dev/stan.git
```
Comment (rok-cesnovar) [inline Jenkinsfile:368]: You only need the Math library.

File: Jenkinsfile [hunk 2]
```diff
@@ -342,6 +355,40 @@ pipeline {
...
+                    cd stan && make math-revert && make clean-all && git clean -xffd && git submodule update --init --recursive
+                """
+
+                writeFile(file: "make/local", text: "CXX=${env.CXX} -Werror ")
+
+                unstash 'ubuntu-exe'
+                // Where to copy the binary ?
```
Comment (rok-cesnovar) [inline Jenkinsfile:375]: To folder-where-math-is-cloned/bin/stanc

File: Jenkinsfile [hunk 3]
```diff
@@ -342,6 +355,40 @@ pipeline {
...
+
+                script {
+                    dir("stan/lib/stan_math/") {
+                        sh "echo O=0 > make/local"
+                        withEnv(['PATH+TBB=./lib/tbb']) {
+                            try { sh "./runTests.py -j${env.PARALLEL} test/expressions" }
+                            finally { junit 'test/**/*.xml' }
+                        }
+                        withEnv(['PATH+TBB=./lib/tbb']) {
+                            sh "python ./test/expressions/test_expression_testing_framework.py"
```
Comment (rok-cesnovar) [inline Jenkinsfile:385]: This was not explicitly stated in the issue but this second test is not required. This is a test of the "expression test framework" so this doesnt have to be run here as well.

**Review by rok-cesnovar — COMMENTED** [review]
First round of comments. I am going to push a change to the file so we can test this.

File: Jenkinsfile (hunk 1 above)
Comment (rok-cesnovar) [reply Jenkinsfile:368]: So only clone math to somewhere and then move the bin/stanc and that should be it.

**Comment by serban-nicusor-toptal** [comment]
Sweet! Give me a moment to fix all the comments.

**Comment by serban-nicusor-toptal** [comment]
I think we're good now, thanks for the quick answer on this.

File: Jenkinsfile [hunk 4]
```diff
@@ -342,6 +355,37 @@ pipeline {
             }
 
         }
+        stage('Math functions expressions test') {
```
Comment (rok-cesnovar) [inline Jenkinsfile:358]: I would move this in the "Cmdstan tests" parallel stage and rename that to CmdStan & Math tests. Other than that looks good, assuming tests pass.

**Comment by rok-cesnovar** [comment]
Looks good now, the error its throwing is something we need to look at in Math.

**Review by rok-cesnovar — APPROVED** [review]

**Pull request merged**